gateway/ln-gateway/src/lib.rs

```diff
@@ -329,14 +375,75 @@ impl Gateway {
         &self,
         BackupPayload { federation_id }: BackupPayload,
     ) -> Result<()> {
-        self.select_actor(federation_id).await?.backup().await
+        self.select_actor(federation_id)
+            .await?
+            .read()
+            .await
+            .backup()
+            .await
     }
 
     async fn handle_restore_msg(
         &self,
         RestorePayload { federation_id }: RestorePayload,
     ) -> Result<()> {
-        self.select_actor(federation_id).await?.restore().await
+        self.select_actor(federation_id)
+            .await?
+            .read()
+            .await
+            .restore()
+            .await
+    }
+
+    async fn handle_lightning_reconnect(
+        &mut self,
+        payload: LightningReconnectPayload,
+    ) -> Result<()> {
+        let LightningReconnectPayload { node_type } = payload;
+
+        let actors = self.actors.lock().await;
+
+        // Stop all threads that are listening for HTLCs
+        tracing::info!("Stopping all HTLC subscription threads.");
+        for (_, actor) in actors.iter() {
+            actor.write().await.stop_subscribing_htlcs().await?;
+        }
+
+        // Disconnect the lightning connection, then reconnect it
+        self.lnrpc.write().await.disconnect().await?;
+
+        let new_lnrpc: Arc<RwLock<dyn ILnRpcClient>> = match node_type {
+            Some(Mode::Cln { cln_extension_addr }) => Arc::new(RwLock::new(
+                NetworkLnRpcClient::new(cln_extension_addr).await?,
+            )),
+            Some(Mode::Lnd {
+                lnd_rpc_addr,
+                lnd_tls_cert,
+                lnd_macaroon,
+            }) => Arc::new(RwLock::new(
+                GatewayLndClient::new(
+                    lnd_rpc_addr,
+                    lnd_tls_cert,
+                    lnd_macaroon,
+                    self.task_group.make_subgroup().await,
+                )
+                .await?,
+            )),
+            None => self.lnrpc.clone(),
+        };
+
+        new_lnrpc.write().await.connect().await?;
```

Comment:
This looks like its connecting twice unless you're reusing the existing client (once in the constructor and once here)

Reply:
Fixed this, will probably do more refactoring when I do this: https://github.com/fedimint/fedimint/issues/2069